### DISCUSSION TITLE
How to change the hatch pattern origin point

### DISCUSSION THREAD

**Discussion by subsurfaceiodev** [mozman/ezdxf#769] [Q&A]
Hi there! Sorry for the inconvenience, but we are not able to find a way to specify the HATCH origin point as in https://www.cadlinecommunity.co.uk/hc/en-us/articles/7095892696605-AutoCAD-Tip-Changing-the-hatch-origin. Thanks!

**Comment by mozman** [comment]
The DXF reference isn't very informative about **how** to implemented certain features, so some guesswork is always required.

The `HATCH` entity has some group codes which are described as seed points and I assumed they are an offset to the pattern origin and could implement your requested feature, but that didn't work, the seed points are something different.

The original pattern with origin at (0.5, 0.5):
![pattern0](https://user-images.githubusercontent.com/4230966/200475757-5cb61690-2c2b-4b26-b349-6e576d1279de.png)


Resetting the pattern origin works by resetting the base point of all pattern lines. 
Origin reset to (0, 0):
![pattern1](https://user-images.githubusercontent.com/4230966/200475856-b29aeac1-8c7f-4b73-bd5b-d2b847240817.png)


```Python

import ezdxf
from ezdxf.entities import Hatch, Pattern
from ezdxf.math import Vec2

ORIGIN = (0.5, 0.5)

def reset_pattern_origin(hatch: Hatch, origin: Vec2):
    if isinstance(hatch.pattern, Pattern):
        for pattern_line in hatch.pattern.lines:
            pattern_line.base_point = origin


doc = ezdxf.new()
msp = doc.modelspace()

points = [(0, 0), (10, 0), (10, 10), (0, 10)]
msp.add_lwpolyline(points, close=True)
hatch = msp.add_hatch()
hatch.paths.add_polyline_path(points)
hatch.set_pattern_fill(
    "MyPattern",
    color=7,
    angle=0,
    scale=1.0,
    style=0,  # normal hatching style
    pattern_type=0,  # user-defined
    definition=[
        [0, ORIGIN, (0, 1), []],
        [90, ORIGIN, (1, 0), []],
    ],
)

# setting the seed points has no effect on the pattern origin:
hatch.set_seed_points([(0.75, 0.75), (0.75, 0.75)])

# resetting the origin works:
reset_pattern_origin(hatch, Vec2(0, 0))

doc.set_modelspace_vport(12, center=(5, 5))
doc.saveas("hatch.dxf")

```

This works for this simple pattern because the origins (base point) of the pattern lines are all the same, this is not true for all hatch patterns. A more sophisticated algorithm may shift all origins by the same amount.

```Python
def shift_pattern_origin(hatch: Hatch, offset: Vec2):
    if isinstance(hatch.pattern, Pattern):
        for pattern_line in hatch.pattern.lines:
            pattern_line.base_point += offset
```

Now you just need to know the offset ...

A possible strategy is to reset the origin of the first pattern line and shift the remaining pattern lines by the same amount:

```Python
def reset_pattern_origin_of_first_pattern_line(hatch: Hatch, origin: Vec2):
    if isinstance(hatch.pattern, Pattern) and len(hatch.pattern.lines):
        first_pattern_line = hatch.pattern.lines[0]
        offset = origin - first_pattern_line.base_point
        shift_pattern_origin(hatch, offset)

```

**Reply by subsurfaceiodev** [reply]
That's exactly what we came up with. Thanks!
